contributors/design-proposals/preserve-order-in-strategic-merge-patch.md

```diff
@@ -0,0 +1,422 @@
+as later environment variables may reference earlier environment variables,
+but not the other way around.
+
+An use case is the environment variables. We don't preserve the order which causes
+issue [40373](https://github.com/kubernetes/kubernetes/issues/40373).
+
+## Proposed Change
+
+Changes are all in strategic merge patch package.
+The proposed solution is similar to the solution used for deleting elements from lists of primitives.
+
+Add to the current patch, a directive containing a list of element keys -
+either the patch merge key, or for primitives the value. When applying the patch,
+the server we ensure that the relative ordering of elements matches the directive.
+
+All the items in the server's live list but not in the parallel list will be append to the end of the parallel list.
+The relative order between these appended items are kept.
+If the relative order of live config in the server is different from the order of the parallel list,
+user's patch will always override the order in the server.
+
+The patched list will looks like:
+
+```
+mergingList:
+- parallelListItem1 \
+  ...                |===> items from the parallel list
+- parallelListItemN /
+- otherItem1 \
+  ...         |===> items in the server's list but not in the parallel list
+- otherItemN /
+```
+
+### `$setElementOrder` may contain elements not present in the patch list
+
+The $setElementOrder value may contain elements that are not present in the patch
+but present in the list to be merge to reorder the elements as part of the merge.
+
+Example where A & B have not changed:
+
+Patch:
+
+```yaml
+$setElementOrder/list:
+- A
+- B
+```
+
+Live:
+
+```yaml
+list:
+- B
+- A
+```
+
+Result:
+
+```yaml
+list:
+- A
+- B
+```
+
+### When the list to be merged contains elements not found in `$setElementOrder`
+
+If the list to be merged contains elements not found in $setElementOrder,
+they will come after all elements defined in $setElementOrder, but keep their relative ordering.
+
+Example where A & B have been changed:
+
+```yaml
+$setElementOrder/list:
+- A
+- B
+list:
+- A
+- B
+```
+
+```yaml
+list:
+- C
+- B
+- D
+- A
+- E
+```
+
+Result:
+
+```yaml
+list:
+- A
+- B
+- C
+- D
+- E
+```
+
+### When `$setElementOrder` contains elements not found in the list to be merged
+
+If `$setElementOrder` contains elements not found in the list to be merged,
+the elements that are not found will be ignored instead of failing the request.
+
+Patch:
+```yaml
+$setElementOrder/list:
+- C
+- A
+- B
+list:
+- A
+- B
+```
+
+Live:
+```yaml
+list:
+- A
+- B
+```
+
+Result:
+
+```yaml
+list:
+- A
+- B
+```
+
+## Version Skew and Backwards Compatibility
+
+The new version patch is always a superset of the old version patch.
+The new patch has one additional parallel list which will be dropped by the old server.
+
+The logic for old version patch is still kept,
+so the patch from any arbitrary old clients used to work before will continue to function.
```

Comment:
> so the patch from any arbitrary old clients used to work before will continue to function.

The new directive is optional.  Patch requests without the directive will retain the behavior from previous releases.  This will ensure backward compatibility with old clients.